### DISCUSSION TITLE
[Feature] Specify sensitive config values via environment variables

### DISCUSSION THREAD

**Discussion by vanchaxy** [immich-app/immich#14815] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Currently, config can be provided via UI or json file. 

Some config values, like `oauth.clientSecret` or `smtp.password` are sensitive. It would be nice to provide these values using environment variables. That will be useful for declarative Immich setup using gitops.

Possible implementations:
1. Allow templating inside immich.json. E.g.
```
"oauth": {
    ...,
   "clientSecret": "${IMMICH_OAUTH_CLIENT_SECRET}",
    ...,
},
```
Immich then should replace clientSecret with the value of `IMMICH_OAUTH_CLIENT_SECRET` env.

2. Allow keys overriding using envs. E.g. `IMMICH_CONFIG_oauth_clientSecret` env will override `oauth.clientSecret` key.

The current workaround is templating externally, e.g. with init container before Immich starts or using secrets manager templating capabilities.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
The workarounds you suggest are already perfectly serviceable approaches, no?

**Reply by vanchaxy** [reply]
I wouldn't call it perfect as it adds one more extra step/container in the deployment, so one more step can fail and need support. Not all secret managers support templating, so it also depends on your setup. I'd like to see first-class secrets support by Immich itself.

**Reply by tecosaur** [reply]
It can also be quite awkward, for example I'm currently trying to deploy immich via NixOS and I can't use the `services.immich.settings` method because templating this is incredibly awkward. I could start encrypting the file in Git, but it would be vastly preferable if something like `clientSecretFile` could be used instead.

**Comment by bo0tzz** [comment]
We discussed this in the team and decided against adding support in Immich, since there are already plenty of good approaches to storing config as a secret or templating it out with things like `envsubst`.

**Reply by tecosaur** [reply]
For what it's worth, I don't think either appy well with a NixOS immich deployment, which looks something like this:

```nix
services.immich = {
  enable = true;
  settings = {
    oauth = {
      clientSecret = "bad practices";
      ...
    };
    notifications.smtp = {
      password = "hard to template";
      ...
    };
  };
  ...
}
```

I also find it notable that all of the other apps I'm self-hosting (or considering): Authelia, Forgejo/Gieta, Mealie, Headscale, Vikunja, LLDAP, SFTPGo, and Caddy support the use of secret files for sensitive configuration.

From what I've seen, "just use envsubst etc." isn't a good solution.

**Reply by bo0tzz** [reply]
Even nix has plenty of solutions for this: https://nixos.wiki/wiki/Comparison_of_secret_managing_schemes

**Reply by tecosaur** [reply]
Yep, there are plenty of ways of encrypting files with Nix, but that's not the situation here — we have secret values within a nix expression, if you use the Immich module.

More generally, I think there's good reason that all the tools I mentioned support secret files.

**Reply by vanchaxy** [reply]
@bo0tzz Thanks for the answer. Can you please clarify how feature requests for Immich work? 

Does a closed issue mean that the community can't vote on a feature and the team won't accept external contributions for it?

I can see other options for this without adding support to Immich:
a) support in the official helm chart (e.g. envsubst as an optional init container);
b) adding official documentation about passing secrets to config; 
c) adding a community guide for it.

I can contribute to any of these. What option is acceptable for the team?

**Reply by bo0tzz** [reply]
A closed request means we're not planning to add official support for this. Of course you're still free to discuss in here etc. Note that at this stage we hardly consider feature votes in any case, as our roadmap is long and already pretty clear.

The first two options would still amount to some degree of official support, but a community guide is always welcome.

**Reply by tecosaur** [reply]
For the record, this is what I've ended up doing: https://code.tecosaur.net/tec/golgi/src/commit/e78515dd19d4e70de85ad6939834a4d1e426eada/modules/immich.nix#L83-L94

It's a bit of a hack, but it works.

**Reply by Eoghanmc22** [reply]
I dont understand why passing via env vars is supported for database and redis credentials but not for the OAUTH secrets

**Reply by bo0tzz** [reply]
Because those values are at different levels. DB and redis are required for Immich to start up, OAuth isn't.

**Reply by RynoM** [reply]
I don't see how the argument "because one is required and one is not", leads to the conclusion it should not be supported. Why not support the same method for anything that might be considered sensitive or a secret?

**Reply by Edward13ruf** [reply]
I'll add that I also came up against this and found it very confusing, and a little concerning, that we are treating this particular sensitive value differently to the rest. I appreciate the implementation effort involved is different, but having these values stored in config files rather than environment values goes against generally accepted best security practice and is out of sync with the other secure variables defined in this project which makes it hard for new users trying to find a solution, to say nothing of the complexity of adjusting helm deployments to add innit containers that map sensitive values into the config file. Can we please reconsider so that all sensitive variables are exposed in the same way for users please? Even with the templated fix that others have suggested (which is not straightforward to adjust your helm deployment to include), I'm not a fan of having my secret values held in anything other than k8s secrets and pod memory.

**Comment by diogotcorreia** [comment]
If anyone is trying to figure out how to do this with Nix, I did a workaround here:
https://github.com/diogotcorreia/dotfiles/commit/d6ff49be843d93364ddf524648fde7e722d3bf0d

Proper support from Immich side would still be very much desirable :)
Thanks for the awesome project!

**Reply by dotlambda** [reply]
The NixOS module now has support for specifying sectrets: https://github.com/NixOS/nixpkgs/pull/453185

**Comment by Deliganli** [comment]
If you are using sops with nixos, I have an example how to do it as well;
https://gist.github.com/Deliganli/554dcc0d05fbd859fb768d2ed2c717c7

Downside of this is, we circumvent code in upstream nixpkgs repo. There may be some smart useful bits, where it is looking to `config.services.immich.settings` and doing things depending on the values. Since we are leaving it as `null` and sideloading our own config, it can't be used.

There is no such functionality in the upstream yet, but there may be in the future.

**Comment by cterence** [comment]
Here's a [way](https://github.com/cterence/homelab-gitops/commit/f5689d93b7ed211751bf54aeb148be080a682ee8) to solve the problem with Kubernetes. I use an init container to run envsubst on the config file from the readonly configmap. It outputs the result in an emptyDir volume mounted by both init and main containers. The "templated" configmap is the one I provide to immich main container with the IMMICH_CONFIG_FILE env var.

**Reply by Lau-r-a** [reply]
Funny the maintainers call this a "already perfectly serviceable approaches". Thats hardly the case...

**Reply by cterence** [reply]
Yeah... I had to scratch my head for a while to get this one working.

**Reply by 2o1o0** [reply]
Same here, it is the only chart (40 deployed right now) where I cannot reference a k8s secret directly from values, which is a bit annoying. I've checked your solution @cterence (btw well done on the argocd template!) but it would make life so much easier to have this single secret to be fed from an env var.

**Reply by ojsef39** [reply]
Thanks for showing your solution. I was shocked there is no official way to do this, only ugly workarounds.
It kind of makes me rethink how seriously security is being taken, but fine, my family prefers immich over nextcloud, so what should I do? xD

Just one note the envsubst image you (@cterence) use wasn't updated for 7 years. 
I'm not sure if I would recommend using that, probably a better approach would probably be:

```
image:
  repository: alpine
  tag: "3.23.2"
command:
  - sh
  - -c
  - |
    apk add --no-cache gettext
    envsubst < /config-template/immich-config.yaml > /config/immich-config.yaml

```
Or packaging the image yourself.

Just something you can trip over unnecessarily because there is no native way to do this…
